### PR TITLE
Tenancy: Implement __FALLBACK tenant and __FALLBACK privilege

### PR DESCRIPTION
**Pull request by lipingxue**
This change solved two problems in this issue https://github.com/vmware/docker-volume-vsphere/issues/621.
- support __FALLBACK tenant
- support __FALLBACK privilege
The remain problem "support default_datastore" will be addressed in my next PR. In my next PR, I plan to  address "support default_datastore" and privilege changes (https://github.com/vmware/docker-volume-vsphere/issues/709)

This change also include necessary unit test to support __FALLBACK tenant and __FALLBACK privilege.

### CODE REVIEW

**Comment by lipingxue** [comment]
@msterin @shaominchen

File: esx_service/utils/error_code.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 VM_NOT_BELONG_TO_TENANT = "VM {0} does not belong to any tenant"
 TENANT_NOT_EXIST = "Tenant {0} does not exist"
 TENANT_ALREADY_EXIST = "Tenant {0} already exists"
+TENANT_CREATE_FAILED = "Tenant {0} create failed with err:{1}"
+TENANT_SET_ACCESS_PRIVILEGES_FAILED = "Tenant {0} set access privileges on datastore {1} failed with {2}"
```
Comment (shaominchen) [inline esx_service/utils/error_code.py]: Nit: Let's keep the error messages consistent: either "failed with {}" or "failed with error: {}"

File: esx_service/vmdk_ops.py [hunk 2]
```diff
@@ -252,6 +252,7 @@ def cloneVMDK(vm_name, vmdk_path, opts={}, vm_uuid=None, vm_datastore=None):
         return err("Failed to initialize source volume path {0}: {1}".format(src_path, errMsg))
 
     src_vmdk_path = vmdk_utils.get_vmdk_path(src_path, src_volume)
+    logging.debug("cloneVMDK: src_path=%s, src_volume=%s, src_vmdk_path=%s", src_path, src_volume, src_vmdk_path)
```
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: The change in this file doesn't seem to be related to __FALLBACK feature. If you want to include these changes, please add some description in the review.

File: esx_service/utils/auth_data.py [hunk 3]
```diff
@@ -336,6 +366,7 @@ def connect(self):
         
         if need_create_table:
             self.create_tables()
+            self.create_fallback_tenant_and_privileges()
```
Comment (shaominchen) [inline esx_service/utils/auth_data.py]: This doesn't seem to be the right place to create fallbacks. I think we should add a flag to check if need_crreate_fallback. This way, we can handle the "upgrade" more smoothly. Also in the unit test we shouldn't have the logic to create fallbacks.

File: esx_service/vmdk_ops_test.py [hunk 4]
```diff
@@ -685,7 +689,38 @@ class VmdkTenantTestCase(unittest.TestCase):
     datastore_name = None
     datastore_path = None
     
-
+    def create_fallback_tenant_and_privileges(self):
```
Comment (shaominchen) [inline esx_service/vmdk_ops_test.py]: As mentioned in above comment, we should check whether fallbacks need to be created even if table already exists. This way we don't need to do this in unit tests.

**Comment by lipingxue** [comment]
@msterin Please review this change.

File: esx_service/utils/auth.py [hunk 5]
```diff
@@ -30,6 +32,8 @@
 CMD_DETACH = 'detach'
 
 SIZE = 'size'
+FALLBACK_TENANT = '__FALLBACK'
```
Comment (pdhamdhere) [inline esx_service/utils/auth.py]: FALLBACK => _DEFAULT or _SYSTEM?

File: esx_service/utils/auth_data.py [hunk 6]
```diff
@@ -308,6 +309,35 @@ def get_auth_db_path(self):
 
         """
         return AUTH_DB_PATH
+    
+    def create_fallback_tenant_and_privileges(self):
+        # create a __FALLBACK tenant
+        error_info, tenant = self.create_tenant(
+                                           name=auth.FALLBACK_TENANT, 
+                                           description="This is a fallback tenant",
```
Comment (pdhamdhere) [inline esx_service/utils/auth_data.py]: "This is a system _default tenant"?

File: esx_service/utils/auth_data.py [hunk 7]
```diff
@@ -308,6 +309,35 @@ def get_auth_db_path(self):
 
         """
         return AUTH_DB_PATH
+    
+    def create_fallback_tenant_and_privileges(self):
```
Comment (pdhamdhere) [inline esx_service/utils/auth_data.py]: +1 on splitting into 2 separate methods.

**Comment by lipingxue** [comment]
@pdhamdhere 
__FALLBACK tenant and __FALLBACK privilege are created by system. Admin can  remove the __FALLBACK tenant. Admin can also remove or modify __FALLBACK privilege.

File: esx_service/utils/auth.py [hunk 8]
```diff
@@ -84,7 +149,18 @@ def get_tenant(vm_uuid):
             tenant_name = result[0]
             logging.debug("Found tenant_uuid %s, tenant_name %s", tenant_uuid, tenant_name)
 
-    return None, tenant_uuid, tenant_name
+        return None, tenant_uuid, tenant_name
+    else:
+        error_info, tenant_uuid, tenant_name = get_fallback_tenant()
```
Comment (msterin) [inline esx_service/utils/auth.py]: get_fallback_tenant() seems to be doing pretty much the dame as the code above (line 136-151). I think it would be better wrap it in a function get_tenant_from_db(where_clause, key) , and use it with something like.
```python
get_tenant_from_db("id = ?", (tenant_uuid, ))
```

Basically I think copy-n-paste is evil :-)

File: esx_service/utils/auth.py [hunk 9]
```diff
@@ -42,6 +46,67 @@ def get_auth_mgr():
...
+    _auth_mgr = get_auth_mgr()
+    privileges = []
+    logging.debug("get_fallback_privileges")
+    try:
+        cur = _auth_mgr.conn.execute(
+            "SELECT * FROM privileges WHERE datastore = ?",
+            (FALLBACK_DS,)
+            )
+        privileges = cur.fetchone()
+    except sqlite3.Error as e:
```
Comment (msterin) [inline esx_service/utils/auth.py]: It feels the exceptions are handled too often. I think yu should handle them much higher up in the stack. Let's discuss offline

File: esx_service/utils/auth_data_test.py [hunk 10]
```diff
@@ -476,19 +485,19 @@ def test_list_tenants(self):
...
-                                 tenants_list[1].default_privileges[0][auth_data_const.COL_MOUNT_VOLUME],
-                                 tenants_list[1].default_privileges[0][auth_data_const.COL_MAX_VOLUME_SIZE],
-                                 tenants_list[1].default_privileges[0][auth_data_const.COL_USAGE_QUOTA]
+        tenant2_actual_output = [tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_TENANT_ID],
+                                 tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_DATASTORE],
+                                 tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_CREATE_VOLUME],
+                                 tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_DELETE_VOLUME],
+                                 tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_MOUNT_VOLUME],
+                                 tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_MAX_VOLUME_SIZE],
+                                 tenants_list[tenant2_idx].default_privileges[0][auth_data_const.COL_USAGE_QUOTA]
```
Comment (msterin) [inline esx_service/utils/auth_data_test.py]: Nice seeing unit tests in the PR ! 

A suggestion: I think you need to check for fallback tenant and priviledge being created in the DB as a separate test. And I think there should be a test with deleting the fallback and making sure ops from nontenant fails

File: esx_service/utils/vmdk_utils.py [hunk 11]
```diff
@@ -248,6 +248,15 @@ def get_vm_uuid_by_name(vm_name):
     except:
         return None
 
+def get_vm_name_by_uuid(vm_uuid):
+    """ Returns vm_name for given vm_uuid, or None """
+    si = vmdk_ops.get_si()
+    try:
+        vm = [d for d in si.content.rootFolder.childEntity[0].vmFolder.childEntity if d.config.uuid == vm_uuid]
+        return vm[0].config.name
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: can it be more that 1 element in the list ?

File: esx_service/vmdk_ops_test.py [hunk 12]
```diff
@@ -251,6 +251,7 @@ def testPolicy(self):
             err = vmdk_ops.removeVMDK(vmdk_path)
             self.assertEqual(err == None, unit[2], err)
 
+@unittest.skip("Liping remove")
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: did you want to remove it :-) ?

File: esx_service/utils/auth.py [hunk 13]
```diff
@@ -42,6 +46,67 @@ def get_auth_mgr():
...
+            )
+        result = cur.fetchone()
+    except sqlite3.Error as e:
+        error_info = "Error {0} when querying from tenants table for tenant {1}".format(e, FALLBACK_TENANT)
+        logging.error(error_info)
+        return str(e), None, None
+    if result:
+        # found __FALLBACK tenant
+        tenant_uuid = result[0]
+        logging.debug("Found __FALLBACK tenant, tenant_uuid %s, tenant_name %s", tenant_uuid, FALLBACK_TENANT)
```
Comment (lipingxue) [inline esx_service/utils/auth.py]: Done

File: esx_service/utils/auth.py [hunk 14]
```diff
@@ -42,6 +46,67 @@ def get_auth_mgr():
...
+        return None, tenant_uuid, FALLBACK_TENANT
+    else:
+        # cannot find __FALLBACK tenant
+        logging.debug(error_code.TENANT_NOT_EXIST.format(FALLBACK_TENANT))
+        return None, None, None
+
+
+def get_fallback_privileges():
```
Comment (lipingxue) [inline esx_service/utils/auth.py]: Done

File: esx_service/utils/auth.py [hunk 15]
```diff
@@ -42,6 +46,67 @@ def get_auth_mgr():
...
+    privileges = []
+    logging.debug("get_fallback_privileges")
+    try:
+        cur = _auth_mgr.conn.execute(
+            "SELECT * FROM privileges WHERE datastore = ?",
+            (FALLBACK_DS,)
+            )
+        privileges = cur.fetchone()
+    except sqlite3.Error as e:
+        error_info = "Error {0} when querying privileges table for __FALLBACK privilege".format(e)
```
Comment (lipingxue) [inline esx_service/utils/auth.py]: Done.

File: esx_service/utils/auth.py (hunk 14 above)
Comment (lipingxue) [inline esx_service/utils/auth.py]: Yes. This should just be one entry for fallback privilege.  But in this entry, they have different privileges (create, remove, and mount). That is why I use "privileges".

File: esx_service/utils/auth.py [hunk 16]
```diff
@@ -84,7 +149,18 @@ def get_tenant(vm_uuid):
             tenant_name = result[0]
             logging.debug("Found tenant_uuid %s, tenant_name %s", tenant_uuid, tenant_name)
 
-    return None, tenant_uuid, tenant_name
+        return None, tenant_uuid, tenant_name
+    else:
+        error_info, tenant_uuid, tenant_name = get_fallback_tenant()
+        if error_info:
+            return error_info, None, None
+        if not tenant_uuid:
```
Comment (lipingxue) [inline esx_service/utils/auth.py]: I will keep it as is.

File: esx_service/utils/auth_data.py (hunk 7 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: Done.

File: esx_service/utils/auth_data.py (hunk 3 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: fallback has nothing to do with upgrade.
fallback tenant and fallback privilege will be created automatically when table was first created. 
upgrade happens when we detect that there exist a DB file, but we found the version is not the latest, then we should have some code to upgrade the DB table to the latest. Those part of code is not here yet.

Since user can remove the fallback tenant and fallback privilege, in unit test, we have some test to test the behavior when fallback tenant and fallback privileges are removed. And after our test, we want to return back to the status that fallback tenant and fallback privileges are still in the DB table.

File: esx_service/utils/error_code.py [hunk 17]
```diff
@@ -18,6 +18,8 @@
 VM_NOT_BELONG_TO_TENANT = "VM {0} does not belong to any tenant"
 TENANT_NOT_EXIST = "Tenant {0} does not exist"
 TENANT_ALREADY_EXIST = "Tenant {0} already exists"
+TENANT_CREATE_FAILED = "Tenant {0} create failed with err:{1}"
```
Comment (lipingxue) [inline esx_service/utils/error_code.py]: Done.

File: esx_service/vmdk_ops.py (hunk 2 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Yes. This is not related to FALLBACK feature. The reason I add this debug log is I found it is very difficult when I debug FALLBACK feature, the unit test failed in cloneVMDK related test. I think the log is helpful for others to debug if they hit errors in this part.

File: esx_service/vmdk_ops_test.py (hunk 4 above)
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Same as above.

File: esx_service/utils/auth_data.py (hunk 6 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: I will keep it as FALLBACK.

File: esx_service/utils/auth.py [hunk 18]
```diff
@@ -42,6 +46,67 @@ def get_auth_mgr():
...
     return thread_local._auth_mgr
 
+def get_fallback_tenant():
+    """ 
+        Get __FALLBACK tenant by querying the auth DB.
+        VMs which does not belong to any tenant explicitly will 
+        be assigned to __FALLBACK tenant if __FALLBACK tenant exists
+        Return value:
+        -- error_info: return None on success or error info on failure
+        -- tenant_uuid: return __FALLBACK tenant uuid on success,
```
Comment (lipingxue) [inline esx_service/utils/auth.py]: Done.

File: esx_service/utils/auth.py [hunk 19]
```diff
@@ -42,6 +46,67 @@ def get_auth_mgr():
...
+        -- error_info: return None on success or error info on failure
+        -- tenant_uuid: return __FALLBACK tenant uuid on success,  
+           return None on failure or __FALLBACK tenant does not exist
+        -- tenant_name: return __FALLBACK tenant name on success,
+           return None on failure or __FALLBACK tenant does not exist
+    """
+    tenant_uuid = None
+    _auth_mgr = get_auth_mgr()
+    try:
+        cur = _auth_mgr.conn.execute(
```
Comment (lipingxue) [inline esx_service/utils/auth.py]: I am not going to address it in this PR, but will file an issue to track this enhancement.

File: esx_service/utils/auth.py (hunk 9 above)
Comment (lipingxue) [inline esx_service/utils/auth.py]: We have discussed offline.

File: esx_service/utils/auth.py (hunk 8 above)
Comment (lipingxue) [inline esx_service/utils/auth.py]: These two place are different. For normal tenant, from vm_uuid, we query vms table to find the tenant_uuid of the tenant which own this VM. Then we query the tenant table by id to get tenant_name.

For FALLBACK tenant, we know the tenant name is  "__FALLBACK". We then query the tenant table by name to get tenant_uuid.

File: esx_service/utils/auth_data_test.py (hunk 10 above)
Comment (lipingxue) [inline esx_service/utils/auth_data_test.py]: "test_vmdkops_on_fallback_tenant_vm" did this.
1. It test the create, attach, detach, remove command when FALLBACK tenant and FALLBACK privileges are present
2. It then test the vmdk ops after removing FALLBACK tenant and FALLBACK privilege

What do you mean "you need to check for fallback tenant and privileges being created in the DB as a separate test"? 
FALLBACK tenant and privileges are created when table is created.

File: esx_service/utils/auth.py (hunk 13 above)
Comment (shaominchen) [inline esx_service/utils/auth.py]: Seems like it's still the same?

File: esx_service/utils/vmdk_utils.py (hunk 11 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: I don't think so. I think the vm name should be unique?

File: esx_service/utils/auth_data_test.py (hunk 10 above)
Comment (msterin) [inline esx_service/utils/auth_data_test.py]: > What do you mean "you need to check for fallback tenant and privileges being created in the DB as a separate test"?
FALLBACK tenant and privileges are created when table is created.

is there a sanity check that validates it ? (low priority)

File: esx_service/utils/auth.py (hunk 14 above)
Comment (shaominchen) [inline esx_service/utils/auth.py]: I think we should differentiate the term "privilege" and "right". IIUC, in our service the term "privilege" indicates some rights (create, remove, mount, etc.) on one datastore. One tenant can have multiple "privileges" on different datastores, but there will be one and only one privilege (which includes multiple "rights") on one single datastore.

File: esx_service/utils/auth.py (hunk 15 above)
Comment (shaominchen) [inline esx_service/utils/auth.py]: Seems like it's still the same?

File: esx_service/utils/auth.py (hunk 16 above)
Comment (shaominchen) [inline esx_service/utils/auth.py]: Can you please explain a little bit?

File: esx_service/utils/auth_data_test.py (hunk 10 above)
Comment (lipingxue) [inline esx_service/utils/auth_data_test.py]: In the setUp() function of class "VmdkTenantTestCase", I have the code to check whether FALLBACK tenant and FALLBACK privilege are missing. It they are missing,  the code will try to create them. See function "create_fallback_tenant_and_privileges".

File: esx_service/utils/vmdk_utils.py (hunk 11 above)
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: > I think the vm name should be unique?

no they are not

Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: Then how do we handle this? Assume we already has an VM named "vm1", then I try to create another vm also named "vm1". Then how can I tell which VM is the one I just created?

Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: I think in one esx, the vm name should be unique, right?

File: esx_service/utils/auth.py (hunk 16 above)
Comment (lipingxue) [inline esx_service/utils/auth.py]: I think it is different coding style. No major advantage to make the change.

**Comment by lipingxue** [comment]
I have adressed all comments. Please take a look.

File: esx_service/utils/vmdk_utils.py (hunk 11 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: Made change to "create_vm" and "remove_vm" to avoid using functino "find_vm_by_name".

**Comment by lipingxue** [comment]
CI test failed in "VmdkClone" test. I have added more debug info in the change to debug the CI failure. 

Please ignore the change related to that.
All other comments have been addressed. Please review it. 
@msterin @pdhamdhere

File: esx_service/utils/log_config.py [hunk 20]
```diff
@@ -30,7 +30,7 @@
 # since we rely on it to locate log file name after config is loaded.
 LOG_CONFIG_FILE = "/etc/vmware/vmdkops/log_config.json"
 
-LOG_LEVEL_DEFAULT = 'INFO'
+LOG_LEVEL_DEFAULT = 'DEBUG'
```
Comment (msterin) [inline esx_service/utils/log_config.py]: this should be changed back before merge
